File: src/main/java/br/com/corelabs/npsharpfx/backend/services/LiveServerService.java
/**
 * Copyright (c) CoreLabs. Todos os direitos reservados.
 * Licenciado sob os termos da licença Proprietária CoreLabs.
 * Consulte o arquivo LICENSE na raiz do projeto para mais informações.
 */
package br.com.corelabs.npsharpfx.backend.services;

import java.awt.Desktop;
import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.URI;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Locale;
import java.util.Objects;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.function.Consumer;
import java.util.function.Supplier;

import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;

import javafx.application.Platform;
import javafx.scene.control.Alert;
import javafx.stage.Stage;

public final class LiveServerService {

    private static final String HOST = "127.0.0.1";
    private static final int HTML_BASE_PORT = 5500;
    private static final int PHP_BASE_PORT = 8000;

    private final Stage stage;
    private final Supplier<java.io.File> workspaceSupplier;
    private final Consumer<String> statusUpdater;
    private final ExecutorService ioExecutor;

    private HttpServer htmlServer;
    private Path htmlRoot;
    private int htmlPort = -1;

    private Process phpProcess;
    private Path phpRoot;
    private int phpPort = -1;

    public LiveServerService(
            Stage stage,
            Supplier<java.io.File> workspaceSupplier,
            Consumer<String> statusUpdater
    ) {
        this.stage = stage;
        this.workspaceSupplier = Objects.requireNonNull(workspaceSupplier);
        this.statusUpdater = statusUpdater == null ? ignored -> {} : statusUpdater;
        this.ioExecutor = Executors.newSingleThreadExecutor(runnable -> {
            Thread thread = new Thread(runnable, "npsharp-live-server");
            thread.setDaemon(true);
            return thread;
        });
    }

    public void openWithLiveServer(java.io.File file) {
        if (file == null || !file.exists() || !file.isFile()) {
            showError("Live Server", "Arquivo inválido.");
            return;
        }

        java.io.File workspace = workspaceSupplier.get();
        if (workspace == null || !workspace.exists() || !workspace.isDirectory()) {
            showError("Live Server", "Abra uma pasta/workspace antes de usar o Live Server.");
            return;
        }

        String ext = extension(file.getName());

        if ("html".equals(ext) || "htm".equals(ext)) {
            openHtml(file.toPath(), workspace.toPath());
            return;
        }

        if ("php".equals(ext)) {
            openPhp(file.toPath(), workspace.toPath());
            return;
        }

        showError("Live Server", "Live Server suporta HTML e PHP neste momento.");
    }

    public void stopAll() {
        stopHtml();
        stopPhp();
        ioExecutor.shutdownNow();
    }

    public void stopHtml() {
        if (htmlServer != null) {
            htmlServer.stop(0);
            htmlServer = null;
            htmlRoot = null;
            htmlPort = -1;
            updateStatus("Live Server HTML parado");
        }
    }

    public void stopPhp() {
        if (phpProcess != null) {
            phpProcess.destroy();
            try {
                if (!phpProcess.waitFor(2, java.util.concurrent.TimeUnit.SECONDS)) {
                    phpProcess.destroyForcibly();
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                phpProcess.destroyForcibly();
            }
            phpProcess = null;
            phpRoot = null;
            phpPort = -1;
            updateStatus("Live Server PHP parado");
        }
    }

    private void openHtml(Path file, Path workspace) {
        ioExecutor.submit(() -> {
            try {
                Path root = workspace.toRealPath().normalize();
                Path target = file.toRealPath().normalize();

                if (!target.startsWith(root)) {
                    showError("Live Server", "Arquivo fora da pasta do projeto.");
                    return;
                }

                if (htmlServer == null || htmlRoot == null || !htmlRoot.equals(root)) {
                    stopHtml();
                    htmlPort = findFreePort(HTML_BASE_PORT);
                    htmlRoot = root;

                    htmlServer = HttpServer.create(new InetSocketAddress(HOST, htmlPort), 0);
                    htmlServer.createContext("/", exchange -> handleStaticRequest(exchange, htmlRoot));
                    htmlServer.setExecutor(Executors.newCachedThreadPool(runnable -> {
                        Thread thread = new Thread(runnable, "npsharp-live-html-client");
                        thread.setDaemon(true);
                        return thread;
                    }));
                    htmlServer.start();
                }

                String url = buildUrl(htmlPort, root, target);
                openBrowser(url);
                updateStatus("Live Server HTML: " + url);
            } catch (Exception e) {
                showError("Live Server HTML", firstLine(e.getMessage()));
            }
        });
    }

    private void openPhp(Path file, Path workspace) {
        ioExecutor.submit(() -> {
            try {
                Path root = workspace.toRealPath().normalize();
                Path target = file.toRealPath().normalize();

                if (!target.startsWith(root)) {
                    showError("Live Server", "Arquivo fora da pasta do projeto.");
                    return;
                }

                if (!isCommandAvailable("php")) {
                    showError("Live Server PHP", "PHP não encontrado no PATH. Instale PHP ou configure o caminho do executável.");
                    return;
                }

                if (phpProcess == null || !phpProcess.isAlive() || phpRoot == null || !phpRoot.equals(root)) {
                    stopPhp();

                    phpPort = findFreePort(PHP_BASE_PORT);
                    phpRoot = root;

                    ProcessBuilder builder = new ProcessBuilder(
                            "php",
                            "-S",
                            HOST + ":" + phpPort,
                            "-t",
                            root.toString()
                    );

                    builder.redirectErrorStream(true);
                    phpProcess = builder.start();

                    Process processRef = phpProcess;
                    ioExecutor.submit(() -> {
                        try (var reader = new java.io.BufferedReader(
                                new java.io.InputStreamReader(processRef.getInputStream(), StandardCharsets.UTF_8))) {
                            String line;
                            while ((line = reader.readLine()) != null) {
                                System.out.println("[php-live-server] " + line);
                            }
                        } catch (IOException ignored) {
                            // processo encerrado
                        }
                    });

                    Thread.sleep(350);
                }

                String url = buildUrl(phpPort, root, target);
                openBrowser(url);
                updateStatus("Live Server PHP: " + url);
            } catch (Exception e) {
                showError("Live Server PHP", firstLine(e.getMessage()));
            }
        });
    }

    private void handleStaticRequest(HttpExchange exchange, Path root) throws IOException {
        try {
            String rawPath = exchange.getRequestURI().getPath();
            String decodedPath = URLDecoder.decode(rawPath, StandardCharsets.UTF_8);
            if (decodedPath.startsWith("/")) {
                decodedPath = decodedPath.substring(1);
            }

            Path requested = root.resolve(decodedPath).normalize();

            if (!requested.startsWith(root)) {
                send(exchange, 403, "Forbidden", "text/plain; charset=utf-8");
                return;
            }

            if (Files.isDirectory(requested)) {
                Path index = requested.resolve("index.html");
                if (Files.exists(index) && Files.isRegularFile(index)) {
                    requested = index;
                }
            }

            if (!Files.exists(requested) || !Files.isRegularFile(requested)) {
                send(exchange, 404, "Not Found", "text/plain; charset=utf-8");
                return;
            }

            byte[] data = Files.readAllBytes(requested);
            exchange.getResponseHeaders().set("Content-Type", contentType(requested));
            exchange.sendResponseHeaders(200, data.length);

            try (OutputStream os = exchange.getResponseBody()) {
                os.write(data);
            }
        } catch (Exception e) {
            send(exchange, 500, "Internal Server Error", "text/plain; charset=utf-8");
        }
    }

    private void send(HttpExchange exchange, int code, String text, String contentType) throws IOException {
        byte[] data = text.getBytes(StandardCharsets.UTF_8);
        exchange.getResponseHeaders().set("Content-Type", contentType);
        exchange.sendResponseHeaders(code, data.length);
        try (OutputStream os = exchange.getResponseBody()) {
            os.write(data);
        }
    }

    private String buildUrl(int port, Path root, Path target) {
        String relative = root.relativize(target).toString().replace('\\', '/');
        return "http://" + HOST + ":" + port + "/" + relative;
    }
private void openBrowser(String url) throws Exception {
    if (isLinux()) {
        new ProcessBuilder("xdg-open", url).start();
        return;
    }

    if (Desktop.isDesktopSupported()
            && Desktop.getDesktop().isSupported(Desktop.Action.BROWSE)) {
        Desktop.getDesktop().browse(URI.create(url));
        return;
    }

    updateStatus("Abra no navegador: " + url);
}

private boolean isLinux() {
    String os = System.getProperty("os.name", "").toLowerCase(Locale.ROOT);
    return os.contains("linux");
}

    private int findFreePort(int start) throws IOException {
        for (int port = start; port < start + 200; port++) {
            try (java.net.ServerSocket socket = new java.net.ServerSocket()) {
                socket.setReuseAddress(false);
                socket.bind(new InetSocketAddress(HOST, port));
                return port;
            } catch (IOException ignored) {
                // tenta próxima porta
            }
        }

        throw new IOException("Nenhuma porta livre encontrada a partir de " + start);
    }

    private boolean isCommandAvailable(String command) {
        try {
            Process process = new ProcessBuilder(command, "-v")
                    .redirectErrorStream(true)
                    .start();

            boolean finished = process.waitFor(3, java.util.concurrent.TimeUnit.SECONDS);
            return finished && process.exitValue() == 0;
        } catch (Exception e) {
            return false;
        }
    }

    private String contentType(Path file) {
        String ext = extension(file.getFileName().toString());

        return switch (ext) {
            case "html", "htm" -> "text/html; charset=utf-8";
            case "css" -> "text/css; charset=utf-8";
            case "js", "mjs" -> "application/javascript; charset=utf-8";
            case "json" -> "application/json; charset=utf-8";
            case "svg" -> "image/svg+xml";
            case "png" -> "image/png";
            case "jpg", "jpeg" -> "image/jpeg";
            case "gif" -> "image/gif";
            case "webp" -> "image/webp";
            case "ico" -> "image/x-icon";
            case "txt" -> "text/plain; charset=utf-8";
            default -> "application/octet-stream";
        };
    }

    private String extension(String name) {
        int dot = name == null ? -1 : name.lastIndexOf('.');
        if (dot < 0 || dot == name.length() - 1) {
            return "";
        }
        return name.substring(dot + 1).toLowerCase(Locale.ROOT);
    }

    private void updateStatus(String message) {
        Platform.runLater(() -> statusUpdater.accept(message));
    }

    private void showError(String title, String message) {
        Platform.runLater(() -> {
            Alert alert = new Alert(Alert.AlertType.ERROR);
            alert.initOwner(stage);
            alert.setTitle(title);
            alert.setHeaderText(null);
            alert.setContentText(message == null || message.isBlank() ? "Erro desconhecido." : message);
            alert.showAndWait();
        });
    }

    private String firstLine(String message) {
        if (message == null || message.isBlank()) {
            return "Erro desconhecido.";
        }
        return message.lines().findFirst().orElse(message);
    }
}

File: src/services/node/liveServerService.ts
/*---------------------------------------------------------------------------------------------
- Copyright (c) CorelabsBR. All rights reserved.
- Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
import { shell } from "electron";
import fs from "node:fs/promises";
import http, { type Server } from "node:http";
import net from "node:net";
import path from "node:path";
import { spawn, type ChildProcess } from "node:child_process";
import type { GitOperationResult, LiveServerRequest, LiveServerResult } from "../../shared/types";
import { commandExists } from "./processService";

const HOST = "127.0.0.1";
const HTML_BASE_PORT = 5500;
const PHP_BASE_PORT = 8000;

let htmlServer: Server | undefined;
let htmlRoot = "";
let htmlPort = -1;
let phpProcess: ChildProcess | undefined;
let phpRoot = "";
let phpPort = -1;

export async function openLiveServer(request: LiveServerRequest): Promise<LiveServerResult> {
  const workspace = path.resolve(request.workspace);
  const filePath = path.resolve(request.filePath);
  const extension = path.extname(filePath).toLowerCase();

  if (!isPathInside(filePath, workspace)) {
    return { success: false, output: "Arquivo fora da pasta do projeto." };
  }

  if (extension === ".html" || extension === ".htm") {
    return openHtml(filePath, workspace);
  }

  if (extension === ".php") {
    return openPhp(filePath, workspace);
  }

  return { success: false, output: "Live Server suporta HTML e PHP neste momento." };
}

export async function stopAllLiveServers(): Promise<GitOperationResult> {
  if (htmlServer) {
    htmlServer.close();
    htmlServer = undefined;
    htmlRoot = "";
    htmlPort = -1;
  }
  if (phpProcess) {
    phpProcess.kill();
    phpProcess = undefined;
    phpRoot = "";
    phpPort = -1;
  }
  return { success: true, output: "Live Server parado" };
}

async function openHtml(filePath: string, workspace: string): Promise<LiveServerResult> {
  if (!htmlServer || htmlRoot !== workspace) {
    if (htmlServer) htmlServer.close();
    htmlPort = await findFreePort(HTML_BASE_PORT);
    htmlRoot = workspace;
    htmlServer = http.createServer((req, res) => {
      void handleStaticRequest(req.url ?? "/", htmlRoot, res);
    });
    await new Promise<void>(resolve => htmlServer!.listen(htmlPort, HOST, resolve));
  }

  const url = buildUrl(htmlPort, workspace, filePath);
  await shell.openExternal(url);
  return { success: true, output: `Live Server HTML: ${url}`, url };
}

async function openPhp(filePath: string, workspace: string): Promise<LiveServerResult> {
  const php = await commandExists("php");
  if (!php) {
    return { success: false, output: "PHP não encontrado no PATH. Instale PHP ou configure o caminho do executável." };
  }

  if (!phpProcess || !phpRoot || phpRoot !== workspace || phpProcess.killed) {
    if (phpProcess) phpProcess.kill();
    phpPort = await findFreePort(PHP_BASE_PORT);
    phpRoot = workspace;
    phpProcess = spawn(php, ["-S", `${HOST}:${phpPort}`, "-t", workspace], {
      cwd: workspace,
      shell: false,
      stdio: "ignore",
      windowsHide: true
    });
    await new Promise(resolve => setTimeout(resolve, 350));
  }

  const url = buildUrl(phpPort, workspace, filePath);
  await shell.openExternal(url);
  return { success: true, output: `Live Server PHP: ${url}`, url };
}

async function handleStaticRequest(rawUrl: string, root: string, res: http.ServerResponse): Promise<void> {
  try {
    const decoded = decodeURIComponent(rawUrl.split("?")[0].replace(/^\/+/, ""));
    let requested = path.resolve(root, decoded);
    if (!isPathInside(requested, root)) {
      send(res, 403, "Forbidden", "text/plain; charset=utf-8");
      return;
    }

    const stat = await fs.stat(requested).catch(() => undefined);
    if (stat?.isDirectory()) {
      requested = path.join(requested, "index.html");
    }

    const fileStat = await fs.stat(requested).catch(() => undefined);
    if (!fileStat?.isFile()) {
      send(res, 404, "Not Found", "text/plain; charset=utf-8");
      return;
    }

    const data = await fs.readFile(requested);
    res.writeHead(200, { "Content-Type": contentType(requested) });
    res.end(data);
  } catch (error) {
    console.warn(`[Sharp-OSS liveServer] Failed to serve ${rawUrl}`, error);
    send(res, 500, "Internal Server Error", "text/plain; charset=utf-8");
  }
}

function send(res: http.ServerResponse, code: number, text: string, contentTypeValue: string): void {
  res.writeHead(code, { "Content-Type": contentTypeValue });
  res.end(text);
}

function buildUrl(port: number, root: string, target: string): string {
  const relative = path.relative(root, target).replace(/\\/g, "/");
  return `http://${HOST}:${port}/${encodePathForUrl(relative)}`;
}

function isPathInside(candidate: string, root: string): boolean {
  const relative = path.relative(root, candidate);
  return relative === "" || (!relative.startsWith("..") && !path.isAbsolute(relative));
}

function encodePathForUrl(value: string): string {
  return value.split("/").map(segment => encodeURIComponent(segment)).join("/");
}

async function findFreePort(start: number): Promise<number> {
  for (let port = start; port < start + 200; port++) {
    if (await isPortFree(port)) return port;
  }
  throw new Error(`Nenhuma porta livre encontrada a partir de ${start}`);
}

function isPortFree(port: number): Promise<boolean> {
  return new Promise(resolve => {
    const server = net.createServer();
    server.once("error", () => resolve(false));
    server.once("listening", () => {
      server.close(() => resolve(true));
    });
    server.listen(port, HOST);
  });
}

function contentType(file: string): string {
  switch (path.extname(file).toLowerCase()) {
    case ".html":
    case ".htm":
      return "text/html; charset=utf-8";
    case ".css":
      return "text/css; charset=utf-8";
    case ".js":
    case ".mjs":
      return "application/javascript; charset=utf-8";
    case ".json":
      return "application/json; charset=utf-8";
    case ".svg":
      return "image/svg+xml";
    case ".png":
      return "image/png";
    case ".jpg":
    case ".jpeg":
      return "image/jpeg";
    case ".gif":
      return "image/gif";
    case ".webp":
      return "image/webp";
    case ".ico":
      return "image/x-icon";
    case ".txt":
      return "text/plain; charset=utf-8";
    default:
      return "application/octet-stream";
  }
}
